File: houseRobber2.go
func rob(nums []int) int {
    if len(nums) == 1 {
        return nums[0]
    }

    return max(robLinear(nums[:len(nums)-1]), robLinear(nums[1:]))
}

func robLinear(houses []int) int {
    prev, curr := 0, 0
    for _, money := range houses {
        prev, curr = curr, max(curr, prev+money)
    }
    return curr
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: houseRobber2.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        def rob_linear(houses: List[int]) -> int:
            prev = curr = 0
            for money in houses:
                prev, curr = curr, max(curr, prev + money)
            return curr

        n = len(nums)
        if n == 1:
            return nums[0]
        
        return max(rob_linear(nums[:-1]), rob_linear(nums[1:]))
